File: fornt-end/frontend-hostelmanagement/src/controller/UserController.js
import axios from "axios";
import { ACCOUNT_API } from "../api/api";
import { useDataByUrl } from "../fetch/FetchData";

export const useGetAllUserExceptRole = (page, size, search, sort, direction, role) => {
    return useDataByUrl({
        url: ACCOUNT_API.GET_ALL_USER_EXCEPT_ROLE(page, size, search, sort, direction, role),
        key: `user_except_role_${page}_${size}_${search}_${sort}_${direction}_${role}`,
    });
}

export const useGetAllUserByRole = (page, size, search, sort, direction, role) => {
    return useDataByUrl({
        url: ACCOUNT_API.GET_ALL_USER_BY_ROLE(page, size, search, sort, direction, role),
        key: `user_by_role_${page}_${size}_${search}_${sort}_${direction}_${role}`,
    });
}

export const useGetAccountById = (userId) => {
    return useDataByUrl({
        url: ACCOUNT_API.GET_ACCOUNT_BY_ID(userId),
        key: `user_${userId}`,
    });
}

export const createAccount = async (account) => {
    try {
        const response = await axios.post(ACCOUNT_API.CREATE, { ...account });
        return response.data;
    } catch (error) {
        throw error.response?.data || { message: "Create account failed!" };
    }
}

export const updateAccount = async (accountId, account) => {
    try {
        const response = await axios.put(ACCOUNT_API.UPDATE(accountId), { ...account });
        return response.data;
    } catch (error) {
        throw error.response?.data || { message: "Update account failed!" };
    }
}

export const banAccount = async (accountId) => {
    try {
        const response = await axios.put(ACCOUNT_API.BAN_ACCOUNT(accountId));
        return response.data;
    } catch (error) {
        throw error.response?.data || { message: "Ban account failed!" };
    }
}

export const unbanAccount = async (accountId) => {
    try {
        const response = await axios.put(ACCOUNT_API.UNBAN_ACCOUNT(accountId));
        return response.data;
    } catch (error) {
        throw error.response?.data || { message: "Unban account failed!" };
    }
}

File: hostel-management-system/src/main/java/t3h/hostelmanagementsystem/controller/user/UserController.java
package t3h.hostelmanagementsystem.controller.user;

import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.web.PagedModel;
import org.springframework.web.bind.annotation.*;
import t3h.hostelmanagementsystem.dto.request.UserDTO;
import t3h.hostelmanagementsystem.dto.response.ApiResponse;
import t3h.hostelmanagementsystem.service.user.UserService;

import java.util.List;

@RestController
@RequestMapping("/owner")
@CrossOrigin("*")
public class UserController {
    private final UserService userService;

    public UserController(UserService userService) {
        this.userService = userService;
    }

    @GetMapping("/get-all-user/except-role/{role}")
    public ApiResponse<List<UserDTO>> getAllUserExceptRole(@PathVariable String role) {
        ApiResponse<List<UserDTO>> apiResponse = new ApiResponse<>();
        apiResponse.setResult(userService.getAllUserExceptRole(role));
        return apiResponse;
    }

    @GetMapping("/get-all-user/by-role/{role}")
    public ApiResponse<PagedModel<UserDTO>> getAllUserByRole(
            @PathVariable String role,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size,
            @RequestParam(required = false) String search,
            @RequestParam(required = false, defaultValue = "id") String sort,
            @RequestParam(required = false, defaultValue = "ASC") String direction
    ) {
        Sort.Direction sortDirection = Sort.Direction.fromString(direction.toUpperCase());
        Sort sortBy = Sort.by(sortDirection, sort);
        Pageable pageable = PageRequest.of(page, size, sortBy);

        ApiResponse<PagedModel<UserDTO>> apiResponse = new ApiResponse<>();
        Page<UserDTO> pageResult = userService.getAllUserByRoleByPage(role, search, pageable);
        PagedModel<UserDTO> pagedModel = new PagedModel<>(pageResult);
        apiResponse.setResult(pagedModel);
        return apiResponse;
    }

    @PostMapping("/create-account")
    public ApiResponse<UserDTO> createUser(@RequestBody @Valid UserDTO userDTO) {
        ApiResponse<UserDTO> apiResponse = new ApiResponse<>();
        apiResponse.setResult(userService.createUser(userDTO));
        return apiResponse;
    }

    @PutMapping("/update-user/{userId}")
    public ApiResponse<UserDTO> updateUser(@PathVariable Long userId, @RequestBody @Valid UserDTO userDTO) {
        ApiResponse<UserDTO> apiResponse = new ApiResponse<>();
        apiResponse.setResult(userService.updateUser(userId, userDTO));
        return apiResponse;
    }

    @GetMapping("/get-user-by-id/{userId}")
    public ApiResponse<UserDTO> getUserById(@PathVariable Long userId) {
        ApiResponse<UserDTO> apiResponse = new ApiResponse<>();
        apiResponse.setResult(userService.getUserById(userId));
        return apiResponse;
    }

    @PutMapping("/ban-account/{userId}")
    public ApiResponse<UserDTO> banAccount(@PathVariable Long userId) {
        ApiResponse<UserDTO> apiResponse = new ApiResponse<>();
        apiResponse.setResult(userService.banAccount(userId));
        return apiResponse;
    }

    @PutMapping("/unban-account/{userId}")
    public ApiResponse<UserDTO> unBanAccount(@PathVariable Long userId) {
        ApiResponse<UserDTO> apiResponse = new ApiResponse<>();
        apiResponse.setResult(userService.unBanAccount(userId));
        return apiResponse;
    }
}
